### PR TITLE
WIP: Check available services at runtime

### PR DESCRIPTION
**Pull request by mocobeta**

### CODE REVIEW

**Comment by mocobeta** [comment]
@dweiss I suppose it can look like this (?)

**Comment by mocobeta** [comment]
One sad thing about relying on the binary distribution would be, that the whole artifact cannot be resolved (I'm not sure though, there could be modules that are not used in Luke but contain service providers).

**Comment by dweiss** [comment]
What do you mean "the whole artifact"? It's the whole point to separate the actual distribution from tests so that we can include as many (or as few) modules as necessary - you'd just filter the modules you want.

File: lucene/distribution-tests/src/test/org/apache/lucene/distribution/TestModularLayer.java [hunk 1]
```diff
@@ -322,4 +322,61 @@ public void testRuntimeModuleCheckAssertions() throws Exception {
                 // obvious.
                 Map.entry("unmap.supported", "true")));
   }
+
+  @Test
+  public void testRuntimeModuleSPICheckAssertions() throws Exception {
+    String luceneCoreModule = "org.apache.lucene.core";
+    // these modules depends on external libraries that are not contained in the binary
+    // distribution.
```
Comment (dweiss) [inline lucene/distribution-tests/src/test/org/apache/lucene/distribution/TestModularLayer.java:330]: Maybe we should add them if Luke makes use of them? It's an open call.

**Review by dweiss — COMMENTED** [review]
Yep, it's something like that that I had in mind. Perhaps we can move it outside of lucene-core, like Uwe said (to a separate module) -- I'll give it a thought.

**Comment by mocobeta** [comment]
"the whole artifact" might be vague word...

I wanted to check all SPIs that are provided by Lucene in the test for completeness, but there are modules that cannot be resolved with the binary distribution, so I excluded them. (I am not sure Luke covers all SPIs in Lucene.)

If such completeness is not required at all on the distribution test, it is fine.

**Comment by dweiss** [comment]
I think we should get there, @mocobeta . I'm still exploring various alternatives, so thank you for bearing with me!

**Comment by dweiss** [comment]
Sorry, closed accidently. Let me experiment with the tests and module configurations a bit more. I think it should be possible (and trivial) to write some sanity checks for the module-layer so that these tests don't have to reside in Luke. Maybe we should also distribute more dependencies so that Luke can use them directly in the binary distribution. This is the point of making such a distribution after all!

**Comment by mocobeta** [comment]
> I think it should be possible (and trivial) to write some sanity checks for the module-layer so that these tests don't have to reside in Luke.

+1.
For the black-box testing for service providers, I thought we could (temporarily) collect or traverse all artifacts and their all dependencies into one module loader, instead of running the module-level tests upon the binary distribution.

> Maybe we should also distribute more dependencies so that Luke can use them directly in the binary distribution. This is the point of making such a distribution after all!

It would be an option. Meanwhile, I think it'd also be desirable to minimize the binary distribution because current Luke uses a small fraction of the artifacts and it will not be likely to expose all Lucene features in the app...

**Comment by mocobeta** [comment]
> I'm still exploring various alternatives

Sure, this PR is also an incomplete prototype for experiments. I have no problem with making another try or redoing from scratch.

**Pull request closed without merging**